### PR TITLE
Tracing: Various fixes and optimizations

### PR DESCRIPTION
**Pull request by nashif**
- Align all trace points for switching thread in/out on all architectures
- trace mutex/semaphore using dedicated calls
- ztest: set thread name
- trace k_sleep

### CODE REVIEW

File: arch/arc/core/switch.S [hunk 1]
```diff
@@ -132,6 +132,17 @@ _switch_return_from_coop:
 
 	pop_s blink
 #endif /* CONFIG_EXECUTION_BENCHMARKING */
+#ifdef CONFIG_TRACING
```
Comment (andrewboie) [inline arch/arc/core/switch.S:135]: not sure I understand why this change was made.

It seems far better to me to do this in the core kernel, around invocations of `arch_switch()` or `arch_swap()` appropriately.

I really like to avoid doing stuff in arch code if we can do it in the core kernel, x1000 if doing in the arch layer requires assembly!

File: include/arch/arc/v2/irq.h [hunk 2]
```diff
@@ -92,9 +92,6 @@ extern void z_irq_spurious(void *unused);
 
 static inline void arch_isr_direct_header(void)
 {
-#ifdef CONFIG_TRACING
```
Comment (andrewboie) [inline include/arch/arc/v2/irq.h:95]: shouldn't this be changed to `CONFIG_TRACING_ISR` instead of completely removed? "direct" interrupts skip the common interrupt handling code AFAIK

File: kernel/include/kswap.h [hunk 3]
```diff
@@ -117,6 +117,7 @@ static ALWAYS_INLINE unsigned int do_swap(unsigned int key,
 		wait_for_switch(new_thread);
 		arch_switch(new_thread->switch_handle,
 			     &old_thread->switch_handle);
+		sys_trace_thread_switched_in();
```
Comment (andrewboie) [inline kernel/include/kswap.h]: commit message mentions x86_64, but this is in common code and applies to all arches using `CONFIG_SWITCH`

File: include/tracing/tracing.h [hunk 4]
```diff
@@ -132,6 +132,41 @@
  */
 #define sys_trace_idle()
 
+/**
+ * @brief Trace initialisation of a Semaphore
+ * @param sem Semaphore object
+ */
+#define sys_trace_semaphore_init(sem)
```
Comment (andrewboie) [inline include/tracing/tracing.h:139]: what do these empty definitions do? is this for when tracing is disabled? I don't think we would need doxygen here.

**Review by tyrantlizard77 — COMMENTED** [review]
Main concern is context switching trace points in arch asm code.
It may be the case that systems that use `arch_swap()` this is unavoidable due to how `arch_swap()` itself gets called from interrupt exit paths, but unless I've missed a detail somewhere, any `arch_switch()` system should be able to instrument this entirely in the core kernel.

File: arch/arc/core/switch.S (hunk 1 above)
Comment (andrewboie) [reply arch/arc/core/switch.S:135]: Did you have to do this because of https://github.com/zephyrproject-rtos/zephyr/issues/20671 ?

File: include/tracing/tracing.h (hunk 4 above)
Comment (nashif) [reply include/tracing/tracing.h:139]: this for when tracing is disabled, doxygen actually belongs here and not in the implementation.

File: include/arch/arc/v2/irq.h (hunk 2 above)
Comment (nashif) [reply include/arch/arc/v2/irq.h:95]: yes

File: arch/arc/core/switch.S (hunk 1 above)
Comment (nashif) [reply arch/arc/core/switch.S]: no, I switched trace point to C but this code was left by mistake. fixed now.

File: arch/arm/core/aarch64/swap_helper.S [hunk 5]
```diff
@@ -35,7 +35,7 @@ GTEXT(z_arm64_context_switch)
 SECTION_FUNC(TEXT, z_arm64_context_switch)
 #ifdef CONFIG_TRACING
 	stp	xzr, x30, [sp, #-16]!
-	bl	sys_trace_thread_switched_in
+	bl	sys_trace_thread_switched_out
```
Comment (ioannisg) [inline arch/arm/core/aarch64/swap_helper.S:38]: This looks like it was a bug.

Comment (ioannisg) [reply arch/arm/core/aarch64/swap_helper.S:38]: @carlocaione @sandeepbrcm ^^

File: include/arch/arc/v2/irq.h [hunk 6]
```diff
@@ -35,6 +35,8 @@ extern void z_arc_firq_stack_set(void);
 extern void arch_irq_enable(unsigned int irq);
 extern void arch_irq_disable(unsigned int irq);
 extern int arch_irq_is_enabled(unsigned int irq);
+extern void sys_trace_isr_enter(void);
```
Comment (ioannisg) [inline include/arch/arc/v2/irq.h:38]: should't these be guarded with the corresponding ifdef?

Comment (nashif) [reply include/arch/arc/v2/irq.h:38]: they can be, sure.

File: arch/arm/core/aarch64/swap_helper.S (hunk 5 above)
Comment (nashif) [reply arch/arm/core/aarch64/swap_helper.S:38]: yes, it was a bug..

Comment (carlocaione) [reply arch/arm/core/aarch64/swap_helper.S:38]: indeed

**Review by ioannisg — APPROVED** [review]

**Review by dcpleung — APPROVED** [review]

File: arch/x86/core/ia32/swap.S [hunk 7]
```diff
@@ -87,6 +87,19 @@ SECTION_FUNC(TEXT, arch_swap)
...
+#if defined(CONFIG_TRACING)
+
+	/* Save these as we are using to keep track of isr and isr_param */
+	pushl	%eax
+	pushl	%edx
+
+	call	sys_trace_thread_switched_out
+
+	popl	%edx
+	popl	%eax
```
Comment (jenmwms) [inline arch/x86/core/ia32/swap.S:101]: Can you please help me understand why we add `push/pop` of `%eax` here? I know it is another register (for x86), and you mention we want to keep `isr` and `isr_param`, which would need 2 registers. How did you know to use `%eax`?

File: arch/x86/core/ia32/swap.S [hunk 8]
```diff
@@ -363,6 +369,17 @@ CROHandlingDone:
 	pushl arch_timing_swap_start
 	popl arch_timing_value_swap_temp
 time_read_not_needed:
+#endif
+#if defined(CONFIG_TRACING)
+
+	/* Save these as we are using to keep track of isr and isr_param */
+	pushl	%eax
+	pushl	%edx
```
Comment (andrewboie) [inline arch/x86/core/ia32/swap.S:377]: we do need to preserve `arch_swap()`'s return value in `%eax`, but why are we pushing/popping `%edx`

Also editorial complaint that we have two different instrumentation points right after one another. What's the difference between `CONFIG_EXECUTION_BENCHMARKING` and `CONFIG_TRACING`? Can we unify?

File: arch/x86/core/ia32/swap.S [hunk 9]
```diff
@@ -363,6 +369,17 @@ CROHandlingDone:
 	pushl arch_timing_swap_start
 	popl arch_timing_value_swap_temp
 time_read_not_needed:
+#endif
+#if defined(CONFIG_TRACING)
+
+	/* Save these as we are using to keep track of isr and isr_param */
```
Comment (andrewboie) [inline arch/x86/core/ia32/swap.S:375]: this comment is wrong.

copypasta?

File: include/arch/arc/v2/irq.h [hunk 10]
```diff
@@ -104,8 +106,8 @@ static inline void arch_isr_direct_footer(int maybe_swap)
 	    z_arc_v2_aux_reg_read(_ARC_V2_AUX_IRQ_HINT)) {
 		z_arc_v2_aux_reg_write(_ARC_V2_AUX_IRQ_HINT, 0);
 	}
-#ifdef CONFIG_TRACING
-	z_sys_trace_isr_exit();
+#ifdef CONFIG_TRACING_ISR
+	sys_trace_isr_exit();
```
Comment (andrewboie) [inline include/arch/arc/v2/irq.h:110]: at least for `sys_trace_thread` and `sys_trace_sem` functions, we have them compile to nothing if tracing is disabled.

can we do the same for ISR, so we don't have to wrap these invocations in ifdefs?

File: include/tracing/tracing.h [hunk 11]
```diff
@@ -131,6 +135,41 @@
...
+
+/**
+ * @brief Trace taking a Semaphore
+ * @param sem Semaphore object
+ */
+#define sys_trace_semaphore_take(sem)
```
Comment (andrewboie) [inline include/tracing/tracing.h:148]: why are these preprocessor `#defines`?

these should be empty inline functions for type safety

never use a macro when an inline function will suffice IMO

File: arch/x86/core/ia32/swap.S (hunk 8 above)
Comment (nashif) [reply arch/x86/core/ia32/swap.S:377]: > Also editorial complaint that we have two different instrumentation points right after one another. What's the difference between `CONFIG_EXECUTION_BENCHMARKING` and `CONFIG_TRACING`? Can we unify? Maybe we can do this in a follow-up PR but it sticks out

not mentioned anywhere here but the idea is to drop CONFIG_EXECUTION_BENCHMARKING completely, been working with Daniel on this already and showing progress. This is just the first step toward this

File: arch/x86/core/ia32/swap.S (hunk 9 above)
Comment (nashif) [reply arch/x86/core/ia32/swap.S:375]: yeah, copy/paste, let me fix that.

File: kernel/include/kswap.h [hunk 12]
```diff
@@ -113,11 +112,11 @@ static ALWAYS_INLINE unsigned int do_swap(unsigned int key,
 			z_smp_release_global_lock(new_thread);
 		}
 #endif
+		sys_trace_thread_switched_out();
 		_current_cpu->current = new_thread;
 		wait_for_switch(new_thread);
 		arch_switch(new_thread->switch_handle,
 			     &old_thread->switch_handle);
-		sys_trace_thread_switched_in();
```
Comment (jenmwms) [inline kernel/include/kswap.h]: I'm curious, was this part of the problem for counting switched in twice? It looks like it, but wanted to ask.

File: arch/x86/core/ia32/swap.S (hunk 7 above)
Comment (nashif) [reply arch/x86/core/ia32/swap.S:101]: > These same complaints apply to the `CONFIG_EXECUTION_BENCHMARKING` block immediately before it that is already in master branch. Which also brings up the question on why we have two instrumentation points right after one another.

yep, this is going away, there will be only one instrumentation point and we will use tracing for measuring timing for the benchmarks we have right now.

File: kernel/include/kswap.h (hunk 12 above)
Comment (nashif) [reply kernel/include/kswap.h]: this is added again in a later patch, they will be sqaushed, but yes, we were tracing both in C and asm.

File: include/tracing/tracing.h [hunk 13]
```diff
@@ -50,16 +50,19 @@
...
 /**
  * @brief Called when a thread is being created
+ * @param thread Thread structure
  */
 #define sys_trace_thread_create(thread)
 
 /**
  * @brief Called when a thread is being aborted
+ * @param thread Thread structure
  *
```
Comment (jenmwms) [inline include/tracing/tracing.h:66]: nit: extra line in comment

File: kernel/sem.c [hunk 14]
```diff
@@ -69,9 +69,9 @@ int z_impl_k_sem_init(struct k_sem *sem, unsigned int initial_count,
 		return -EINVAL;
 	}
 
-	sys_trace_void(SYS_TRACE_ID_SEMA_INIT);
 	sem->count = initial_count;
 	sem->limit = limit;
+	sys_trace_semaphore_init(sem);
```
Comment (jenmwms) [inline kernel/sem.c:74]: I understand the change to `sys_trace_semaphore_init(sem)`. Why did we move it below the `sem-`>` assignments? In this position, does it better represent the timing of the trace at initializing the semaphore (as in the next line z_waitq_init)?

File: subsys/tracing/include/tracing_test.h [hunk 15]
```diff
@@ -31,7 +31,12 @@ void sys_trace_isr_exit_to_scheduler(void);
 void sys_trace_idle(void);
 void sys_trace_void(unsigned int id);
 void sys_trace_end_call(unsigned int id);
-
+void sys_trace_semaphore_init(struct k_sem *sem);
```
Comment (jenmwms) [inline subsys/tracing/include/tracing_test.h:34]: jw - I expected to see `sys_trace_start_call()` here (and ctf_top.c/h), too similar to the sys_trace_end_call, etc.. Looks like it hasn't been there yet but things work. Is it missing? Please correct me if I'm wrong.

File: arch/x86/core/ia32/intstub.S [hunk 16]
```diff
@@ -203,6 +200,12 @@ alreadyOnIntStack:
 	cli			/* disable interrupts again */
 #endif
 
+#if defined(CONFIG_TRACING_ISR)
+	pushl	%eax
+	call	sys_trace_isr_exit
```
Comment (jenmwms) [inline arch/x86/core/ia32/intstub.S:205]: Does this need the `pushl/popl %edx` like above in line 183?

File: scripts/tracing/parse_ctf.py [hunk 17]
```diff
@@ -0,0 +1,151 @@
+#!/usr/bin/env python3
+#
+# Copyright (c) 2020 Intel Corporation.
+#
+# SPDX-License-Identifier: Apache-2.0
+"""
+Script to parse CTF data and print to the screen in a custom and colorful
```
Comment (jenmwms) [inline scripts/tracing/parse_ctf.py:7]: Do I understand the role of this script? This script looks specific to babeltrace. Does it use babeltrace2 to interpret the CTF channel and metadata, and this script works to customize the display using babeltrace APIs?

File: arch/x86/core/ia32/intstub.S [hunk 18]
```diff
@@ -192,6 +179,16 @@ alreadyOnIntStack:
 	pop %eax
 #endif
 
+
+#if defined(CONFIG_TRACING_ISR)
+	/* Save these as we are using to keep track of isr and isr_param */
+	pushl	%eax
```
Comment (andrewboie) [inline arch/x86/core/ia32/intstub.S:185]: I don't think EAX needs to be pushed here, it used to contain the interrupt handler argument but that got pushed onto the stack earlier

The CONFIG_EXECUTION_BENCHMARKING block is also wrong but that's eventually being removed.

The stench of copy-paste is strong here...have to be very careful about doing this in assembly as there is no compiler to fix redundancies

File: arch/x86/core/ia32/intstub.S (hunk 16 above)
Comment (andrewboie) [reply arch/x86/core/ia32/intstub.S:205]: no I don't believe so, it's unused after the call statement, we even zero it on line 211 if X2APIC is active

but we also don't need to push/pop EAX here either, note how it is zeroed on line 209

File: scripts/tracing/parse_ctf.py (hunk 17 above)
Comment (nashif) [reply scripts/tracing/parse_ctf.py:7]: this is just a helper script to read the generated data on the command line.

File: subsys/tracing/include/tracing_test.h (hunk 15 above)
Comment (nashif) [reply subsys/tracing/include/tracing_test.h:34]: yeah, looks like sys_trace_start_call is missing in the test variant, this is just a test, no harm :)

File: kernel/sem.c (hunk 14 above)
Comment (nashif) [reply kernel/sem.c:74]: > Why did we move it below the `sem-`>` assignments?

to get the values the semaphore was initiated with, this could be part of the tracing later.

File: include/arch/arc/v2/irq.h (hunk 10 above)
Comment (nashif) [reply include/arch/arc/v2/irq.h:112]: yeah, i dont like the too many ifdefs either, this will move the ifdef into the function however, which I think is fine..

File: include/tracing/tracing.h (hunk 11 above)
Comment (nashif) [reply include/tracing/tracing.h:148]: if I remember correctly we used to have them inline initially or at least in the  original PR from 2-3 years ago, some users however wanted them to be defines to make it easy to integrate with existing tracing formats. There was some other complication I do not recall, need to check and will revisit, in any case, this would need to happen in a different PR.

**Review by jenmwms — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Review by ruuddw — APPROVED** [review]

**Review by abrodkin — APPROVED** [review]
LGTM
